server/src/internalClusterTest/java/org/elasticsearch/search/ccs/CrossClusterIT.java

```diff
@@ -0,0 +1,562 @@
+                .state()
+                .nodes()
+                .stream()
+                .map(DiscoveryNode::getName)
+                .filter(nodeName -> nodeWithoutRemoteClusterClientRole.equals(nodeName) == false)
+                .filter(nodeName -> nodeName.equals(pureDataNode) == false)
+                .toList()
+        );
+
+        final SearchResponse resp = localCluster.client(nodeWithRemoteClusterClientRole)
+            .prepareSearch("demo", "cluster_a:prod")
+            .setQuery(new MatchAllQueryBuilder())
+            .setAllowPartialSearchResults(false)
+            .setSize(1000)
+            .get();
+        assertHitCount(resp, demoDocs + prodDocs);
+    }
+
+    public void testProxyConnectionDisconnect() throws Exception {
+        assertAcked(client(LOCAL_CLUSTER).admin().indices().prepareCreate("demo"));
+        indexDocs(client(LOCAL_CLUSTER), "demo");
+        final String remoteNode = cluster("cluster_a").startDataOnlyNode();
+        assertAcked(
+            client("cluster_a").admin()
+                .indices()
+                .prepareCreate("prod")
+                .setSettings(
+                    Settings.builder()
+                        .put("index.routing.allocation.require._name", remoteNode)
+                        .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                        .build()
+                )
+        );
+        indexDocs(client("cluster_a"), "prod");
+        SearchListenerPlugin.blockQueryPhase();
+        try {
+            PlainActionFuture<SearchResponse> future = new PlainActionFuture<>();
+            SearchRequest searchRequest = new SearchRequest("demo", "cluster_a:prod");
+            searchRequest.allowPartialSearchResults(false);
+            searchRequest.setCcsMinimizeRoundtrips(false);
+            searchRequest.source(new SearchSourceBuilder().query(new MatchAllQueryBuilder()).size(1000));
+            client(LOCAL_CLUSTER).search(searchRequest, future);
+            SearchListenerPlugin.waitSearchStarted();
+            disconnectFromRemoteClusters();
+            // Cancellable tasks on the remote cluster should be cancelled
+            assertBusy(() -> {
+                final Iterable<TransportService> transportServices = cluster("cluster_a").getInstances(TransportService.class);
+                for (TransportService transportService : transportServices) {
+                    Collection<CancellableTask> cancellableTasks = transportService.getTaskManager().getCancellableTasks().values();
+                    for (CancellableTask cancellableTask : cancellableTasks) {
+                        if (TransportActionProxy.isProxyAction(cancellableTask.getAction())) {
+                            assertTrue(cancellableTask.getDescription(), cancellableTask.isCancelled());
+                        }
+                    }
+                }
+            });
+            assertBusy(() -> assertTrue(future.isDone()));
```

Comment:
can we not call future.get()? looks odd to assertBusy, although it achieves the same end result I guess.

Reply:
Yeah, I can change it, but I didn't write this code. I just moved it to a new class :-) My new code is in CrossClusterSearchIT. That said I often do the same thing there because in later versions of the test (next ticket) calling future.get throws an (expected) Exception so the pattern I used is just to wait until it's done and then call get. But for tests where an Exception is never expected your suggestion is a good one.